### PR TITLE
fix: clean up data handling in TFKerasTrial [DET-4300]

### PR DESCRIPTION
**Pull request by rb-determined-ai**
## Description

Stop managing OrderedEnqueuers; model.fit() will do that.

Also support deterministic shuffling for Keras Sequence objects.

Rely on new callback support for num_inputs.

Delete the now-unneeded input manager objects.

## Test Plan

Deleted irrelevant tests, updated others, added new ones.

## Checklist

- [ ] Release notes should be added as a separate file under `docs/release-notes/`.

### CODE REVIEW

File: harness/determined/keras/_tf_keras_trial.py [hunk 1]
```diff
@@ -540,37 +531,70 @@ def run(self) -> None:
...
-            validation_data,
-            validation_steps,
-        ) = self._validation_input_manager.get_validation_input_and_num_batches()
+        validation_data = self.validation_data
+
+        if isinstance(validation_data, keras.SequenceAdapter):
+            # Ignore these settings and use the same settings as for the fit call.
+            validation_data = validation_data.sequence
+
+        if isinstance(validation_data, tf.keras.utils.Sequence):
```
Comment (aaron276h) [inline harness/determined/keras/_tf_keras_trial.py:581]: question: what happens when the sequence completes an epoch?

File: harness/determined/keras/_tf_keras_context.py [hunk 2]
```diff
@@ -35,6 +35,33 @@ def __init__(self, env: det.EnvContext, hvd_config: horovod.HorovodContext):
...
 
+        # The following attributes may be configured via _configure_fit().  Defaults match the
+        # normal keras.fit() defaults.
+        self._fit_verbose = True
+        self._fit_shuffle = True
+        self._fit_workers = 1
+        self._fit_use_multiprocessing = False
+        self._fit_max_queue_size = 10
+
+    def _configure_fit(
```
Comment (aaron276h) [inline harness/determined/keras/_tf_keras_context.py:46]: question: is this just for native ?

**Review by aaron276h — COMMENTED** [review]
Left a couple questions, I haven't yet done a detailed pass on `_DeterminedSequenceWrapper`.

File: harness/determined/keras/_tf_keras_context.py (hunk 2 above)
Comment (rb-determined-ai) [reply harness/determined/keras/_tf_keras_context.py:46]: no, this is the public api improvement I mentioned in ml-ag

File: harness/determined/keras/_data.py [hunk 3]
```diff
@@ -123,17 +144,125 @@ def __getitem__(
...
     def __len__(self):  # type: ignore
-        return len(self._sequence)
+        return self._len
 
     def __getitem__(self, index):  # type: ignore
-        index = (index + self._batch_offset) % len(self)
-        return self._sequence[index]
+        return self._sequence[self._indices[index + self._batch_offset]]
+
+    def on_epoch_end(self) -> None:
```
[the remark this answers is not in the document]
Comment (rb-determined-ai) [reply harness/determined/keras/_data.py:254]: yes, it's called by the ordered enqueuer.

File: harness/determined/keras/_tf_keras_trial.py [hunk 4]
```diff
@@ -540,37 +531,70 @@ def run(self) -> None:
...
+                shard_size=self.context.distributed.get_size(),
+                training=False,
+            )
 
         metrics_values = self.model.evaluate(
             validation_data,
-            steps=validation_steps,
             verbose=0,
             callbacks=self.callback_list,
+            use_multiprocessing=self.context._fit_use_multiprocessing,
```
[the remark this answers is not in the document]
Comment (rb-determined-ai) [reply harness/determined/keras/_tf_keras_trial.py:593]: this is exactly what keras does, so yes.

File: harness/determined/keras/_tf_keras_inputs.py [hunk 5]
```diff
@@ -1,307 +0,0 @@
...
-        self,
-        context: Union[keras.TFKerasTrialContext, keras.TFKerasNativeContext],
-        train_config: keras.TFKerasTrainConfig,
-    ) -> None:
-        super().__init__(context=context)
-
-        self._validation_cacheable = self._context.experimental.get_validation_cacheable()
-        self._validation_dataset = train_config.validation_data
-
-        check.true(
```
[the remark this answers is not in the document]
Comment (rb-determined-ai) [reply harness/determined/keras/_tf_keras_inputs.py:116]: oooh, yeah we totally did.  I'll bring them back, good catch.

File: harness/determined/keras/_data.py [hunk 6]
```diff
@@ -169,198 +298,55 @@ def __init__(
...
+    if isinstance(input_data, tf.data.Dataset):
+        return input_data
 
-    elif isinstance(input_data, tuple) and len(input_data) == 3:
-        return input_data[0], input_data[1], input_data[2]
+    if isinstance(input_data, (tf.keras.utils.Sequence, SequenceAdapter)):
+        return input_data
 
-    else:
+    if not isinstance(input_data, tuple) or len(input_data) not in (2, 3):
```
Comment (shiyuann) [inline harness/determined/keras/_data.py:318]: non-blocking: this doesn't seem to be easier to understand nor save lines of code. :)

**Review by shiyuann — COMMENTED** [review]
Nice work in reducing the complexity of handling Keras data! I looked at the high-level refactoring, which looked good! Didn't really went deep into the details since I haven't worked on this for too long.

**Comment by rb-determined-ai** [comment]
closing in favor of #1564
